docs/proposals/read-write-operations-bucket.md

```diff
@@ -0,0 +1,244 @@
+
+We propose defining those set of rules to solve A, B and C goals that will define the object storage format API (and how to operate with those objects):
+
+1. Block once uploaded is never edited (no file within ULID directory is removed, edited or added).
+1. Block is assumed ready to be read *only* after 15 minutes (configured by `syncDelay`) from *creation time*. Block id (`ULID`) is used
+   to determine creation time (`if ulid.Now()-id.Time() < uint64(c.syncDelay/time.Millisecond)`). Block fresher than 15 minutes is assumed
+   still in upload (`Fresh`). This can be overridden by `ignore_delay` field in meta.json.
+1. TSDB `meta.json` file within block *have to be* uploaded as first one.
+1. After 15 minutes (or time configured by `syncDelay`), if there is no fully loadable `meta.json`, we assume block is malformed, partially uploaded or marked for deletion and can be
+   *removed* by compactor. [Compactor change needed]
+1. Overlaps (in external labels and time dimensions) with the same data are allowed and should be handled by readers gracefully. [Compactor change needed]
+1. Adding "replacement" or "overlapped" block causes old blocks to be scheduled for removal when new block is ready by compactor (`HealthyOverlapped` state). [Compactor/Repair change needed]
+1. To schedule delete operation, delete `meta.json` file. All components will exclude this block and compactor will do eventual deletion assuming the block is partially uploaded. [Compactor change needed]
+1. Compactor waits minimum 15m (`deleteDelay`) before deleting the whole `To Delete` block. [Compactor change needed]
+
+### Why?
+
+To deal with A and B, so overall eventual consistency, we will leverage existing feature of metric format (TSDB) which is 
+*block immutability*. In object storage format, a block is a directory named with [`ULID`](https://github.com/oklog/ulid) that contains
+`chunks` dir with one or more chunk files (each not bigger than 512MB), `index` and `meta.json` files. To ensure every component
+has the same state of metrics we create first strict rule:
+
+> 1 . Block once uploaded is never edited (no file within ULID directory is removed, edited or added).
+
+With this rule alone we restrict bucket operations only to:
+* Block upload
+* Block removal 
+
+As those operations are not atomic itself, partial blocks or inconsistent block view may occur for short time (read-after-write consistency).
+To mitigate this we introduce second rule:
+
+ > 2 . Block is assumed ready to be read *only* after 15 minutes (configured by `syncDelay`) from *creation time*. Block id (`ULID`) is used
+to determine creation time (`if ulid.Now()-id.Time() < uint64(c.syncDelay/time.Millisecond)`). Block fresher than 15 minutes is assumed
+still in upload (`Fresh`). This can be overridden by `ignore_delay` field in meta.json.
+
+Readers may attempt to load block's files, but should be aware that those files can be partially uploaded (e.g meta.json to grab external labels).
```

Comment:
Is it that the files are partially uploaded, or that the set of the files uploaded will not be necessarily complete initially? I understand it's the latter, and if so `meta.json` can hold data that is a checksum of the set of files(names), which in turn could let the reader know early if the data is fully available.